OpenRA.Game/Graphics/Util.cs

```diff
@@ -20,12 +20,27 @@ public static class Util
 		// yes, our channel order is nuts.
 		static readonly int[] ChannelMasks = { 2, 1, 0, 3 };
 
-		public static void FastCreateQuad(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv, in float3 size, in float3 tint, float alpha)
+		public static void FastCreateQuad(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv,
+			in float3 size, in float3 tint, float alpha, float rotation = 0f)
 		{
+			var a = o;
 			var b = new float3(o.X + size.X, o.Y, o.Z);
 			var c = new float3(o.X + size.X, o.Y + size.Y, o.Z + size.Z);
 			var d = new float3(o.X, o.Y + size.Y, o.Z + size.Z);
-			FastCreateQuad(vertices, o, b, c, d, r, samplers, paletteTextureIndex, tint, alpha, nv);
+
+			// Rotate sprite if rotation angle is not equal to 0
+			if (rotation != 0f)
+			{
+				var centerPoint = o + (size / 2);
+				var angleSin = (float)Math.Sin(-rotation);
+				var angleCos = (float)Math.Cos(-rotation);
+				a = RotatePoint(o, centerPoint, angleSin, angleCos);
```

Comment:
I suspect that this won't be handling the z coordinate directly.

I would expect the following to behave better, as well as being clearer and more efficient:
```csharp
// Rotate sprite if rotation angle is not equal to 0
if (rotation != 0f)
{
	var center = o + 0.5f * size;
	var angleSin = (float)Math.Sin(-rotation);
	var angleCos = (float)Math.Cos(-rotation);

	// Rotated offset for +/- x with +/- y
	var ra = 0.5f * new float3(
		size.X * angleCos - size.Y * angleSin,
		size.X * angleSin + size.Y * angleCos,
		(size.X * angleSin + size.Y * angleCos) * size.Z / size.Y);

	// Rotated offset for +/- x with -/+ y
	var rb = 0.5f * new float3(
		size.X * angleCos + size.Y * angleSin,
		size.X * angleSin - size.Y * angleCos,
		(size.X * angleSin - size.Y * angleCos) * size.Z / size.Y);

	a = center - ra;
	b = center + rb;
	c = center + ra;
	d = center - rb;
}
else
{
	a = o;
	b = new float3(o.X + size.X, o.Y, o.Z);
	c = new float3(o.X + size.X, o.Y + size.Y, o.Z + size.Z);
	d = new float3(o.X, o.Y + size.Y, o.Z + size.Z);
}
```

Reply:
The Z-coordinate is unaffected by the original code, since the rotation is only on the 2D plane (Yaw). I am not sure how your code above handles the Z-coord. With some brief testing it seems to work the same way as the original code. I can use your code above, however note that I haven't tested it as extensively, and wouldn't be able to reason about its validity (am relying on you for this). If you are okay with these caveats then I can proceed.

Reply:
The point is that the z coordinate *needs to be* affected, since the sprite can be projected onto an inclined plane in z.

Reply:
The issue can be demonstrated by setting `this.rotation = rotation + new WAngle(128);` in the `SpriteRenderable` ctor and then looking at the refinery/weapons factory bibs in TS. You can see visually (and even more clearly using the depth preview) how they are clipped against the terrain in the current PR, but drawn correctly in my suggested version.

So yes please do proceed, and remove the `RotatePoint` method.

Reply:
I have resolved this by using your approach above. I moved it into another function called RotateQuad(), so it can both be reused and also tested. I also replaced the previous test function with a RotateQuadTest() function, though admittedly I did not test the Z coordinate here as I am not entirely sure how that is expected to work.